pkg/mads/v1/reconcile/snapshot_generator.go

```diff
@@ -4,56 +4,131 @@ import (
 	"context"
 
 	envoy_core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	"github.com/pkg/errors"
 
+	"github.com/kumahq/kuma/pkg/core"
 	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
 	core_manager "github.com/kumahq/kuma/pkg/core/resources/manager"
 	core_store "github.com/kumahq/kuma/pkg/core/resources/store"
 	core_xds "github.com/kumahq/kuma/pkg/core/xds"
 	"github.com/kumahq/kuma/pkg/mads/generator"
 	mads_v1_cache "github.com/kumahq/kuma/pkg/mads/v1/cache"
+	meshmetrics_generator "github.com/kumahq/kuma/pkg/mads/v1/generator"
+	"github.com/kumahq/kuma/pkg/plugins/policies/core/matchers"
+	"github.com/kumahq/kuma/pkg/plugins/policies/meshmetric/api/v1alpha1"
 	util_xds_v3 "github.com/kumahq/kuma/pkg/util/xds/v3"
+	"github.com/kumahq/kuma/pkg/xds/cache/mesh"
 )
 
-func NewSnapshotGenerator(resourceManager core_manager.ReadOnlyResourceManager, resourceGenerator generator.ResourceGenerator) util_xds_v3.SnapshotGenerator {
+var log = core.Log.WithName("mads").WithName("v1").WithName("reconcile")
+
+func NewSnapshotGenerator(resourceManager core_manager.ReadOnlyResourceManager, resourceGenerator generator.ResourceGenerator, meshCache *mesh.Cache) util_xds_v3.SnapshotGenerator {
 	return &snapshotGenerator{
 		resourceManager:   resourceManager,
 		resourceGenerator: resourceGenerator,
+		meshCache:         meshCache,
 	}
 }
 
 type snapshotGenerator struct {
 	resourceManager   core_manager.ReadOnlyResourceManager
 	resourceGenerator generator.ResourceGenerator
+	meshCache         *mesh.Cache
 }
 
-func (s *snapshotGenerator) GenerateSnapshot(ctx context.Context, _ *envoy_core.Node) (util_xds_v3.Snapshot, error) {
-	meshes, err := s.getMeshesWithPrometheusEnabled(ctx)
+func (s *snapshotGenerator) GenerateSnapshot(ctx context.Context, node *envoy_core.Node) (util_xds_v3.Snapshot, error) {
+	meshesWithMeshMetrics, err := s.getMeshesWithMeshMetrics(ctx, node.Id)
 	if err != nil {
 		return nil, err
 	}
 
-	dataplanes, err := s.getDataplanes(ctx, meshes)
+	meshes, err := s.getMeshesWithPrometheusEnabled(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	meshGateways, err := s.getMeshGateways(ctx, meshes)
-	if err != nil {
-		return nil, err
+	if len(meshes) > 0 && len(meshesWithMeshMetrics) > 0 {
```

Comment:
Do I understand correctly that if both are configured `MeshMetric` will be used to generate `MonitoringAssignment`? Maybe it is worth to add this info to this log?

Reply:
> Maybe it is worth to add this info to this log? Or don't generate MonitoringAssignment at all?

I think it makes sense to add it to logs. I think we should allow a flow from old to new policies with the least interruptions as possible.